src/async.c

```diff
@@ -79,8 +79,16 @@ mpd_async_new(int fd)
 	async->fd = fd;
 	mpd_error_init(&async->error);
 
-	mpd_buffer_init(&async->input);
-	mpd_buffer_init(&async->output);
+	if (mpd_buffer_init(&async->input) == NULL) {
+		async->output.data = NULL;
```

Comment:
This is an obscure way of fake-initializing the buffer to allow `mpd_async_free()` (the `mpd_buffer` API doesn't document that this is legal). I don't like obscure code like that.
If you refactor this to never fail (like before) and do the first allocation lazily, you can avoid this fragile and cumbersome code.

Reply:
In that case, we would need to do the allocation in `mpd_buffer_room()`; currently, it returns `size_t `which would make the OOM case awkward (return `(size_t)-1` for instance). We could change that function to write the size in a pointer. Would you be okay with any of these solutions?

Reply:
Why would `mpd_buffer_room()` need to do the allocation? It is a simple read-only function which gives information about the buffer state.

Reply:
The idea was to initialize in a single place; i realize now that read functions does not use the `mpd_buffer_room()`
I will update the code to allocate in `mpd_buffer_read()` and `mpd_buffer_write()`

Reply:
Updated code in commit https://github.com/MusicPlayerDaemon/libmpdclient/pull/53/commits/154d1a3ef3aab678c129ea14f8951acb1147b063.
This resulted in an increase of code complexity as the code checks for OOM events in more places.